File: Ch5_TransformAnalysisOfLTISystem/Advanced_EffectOfGroupDelayOnSignals/filter_analysis_imp.m
function filter_analysis_imp(imp_rep,fps,title_)

N = 2^nextpow2(length(imp_rep));
F = fft(imp_rep,N);
freq_step = fps/N;
freq = -fps/2:freq_step:fps/2-freq_step;
freq = freq.*60; %Hz to bpm
freq = freq(N/2+1:end)';

figure;
plot(freq,abs(F(1:N/2)));
xlabel('freq(bpm)');
ylabel('mag');
title(title_);

figure;
Y = angle(F(1:end))*180/pi;

plot(freq,  (Y(1:N/2)));
% Uncomment and think.
% plot(freq,  unwrap(Y(1:N/2)));

xlabel('freq(bpm)');
ylabel('Phase (degree)');
title(title_);

figure;
Y = angle(F(1:end));
Y = -1*diff(Y)/freq_step/2/pi/fps;

plot(freq,  (Y(1:N/2)));
% Uncomment and think.
% plot(freq,  unwrap(Y(1:N/2)));

xlabel('freq(bpm)');
ylabel('Group delay [sec]');
title(title_);


end

File: Ch7_FilterDesignTechniques/Middle_WindowEffect/filter_analysis_imp.py
import numpy as np
import matplotlib.pyplot as plt

def filter_analysis_imp(imp_rep, fps, title_):
    N = int(2 ** np.ceil(np.log2(len(imp_rep))))
    F = np.fft.fft(imp_rep, N)
    freq_step = fps / N
    freq = np.arange(-fps/2, fps/2, freq_step)
    freq = freq * 60  # Hz to bpm
    freq = freq[int(N/2):]

    plt.figure()
    plt.plot(freq, np.abs(F[:N//2]))
    plt.xlabel('freq(bpm)')
    plt.ylabel('mag')
    plt.title(title_)

    plt.figure()
    Y = np.angle(F) * 180 / np.pi
    plt.plot(freq, np.unwrap(Y[:N//2]))
    plt.xlabel('freq(bpm)')
    plt.ylabel('Phase (degree)')
    plt.title(title_)

    plt.figure()
    Y = np.angle(F)
    Y = -1 * np.diff(Y) / freq_step / (2 * np.pi * fps)
    plt.plot(freq[:-1], np.unwrap(Y[:N//2]))
    plt.xlabel('freq(bpm)')
    plt.ylabel('Group delay [sec]')
    plt.title(title_)

    plt.show()
